### ISSUE TITLE
Bug with Table.put()

### ISSUE THREAD

**Issue by Fareley** [dfahlander/Dexie.js#648]
Spent about 5 hours debuggin today, haha. So, imagine the following table declaration:

db.version(1).stores({
  friends: '++appKey, userId, [appKey+userId]'
});

Now, inserting a new entry into friends does create the record, but not the Compound Index record (as part of the compound index is the primary key, which does not have a value yet until after the "add"). This can be seen the easiest when inspecting the IndexedDB with Chrome.

So, what I did to circumvent this was to add and put the same object again when creating it (so the second time around, it does have a primary key already and can create the compound index properly):

var newAppKey = await this.dexieService.friends.add(friend);
await this.dexieService.friends.put(friend, newAppKey);

Now, see, here is the problem: The code above crashes (with one of those zone invalid paramter stuff errors that really say nothing), because I use a "put" on a table with an autoincrement while providing the optional parameter of the primary key as well as the object having a primary key already (due to the "add" above). My issue with this is that the optional parameter should not crash the entire thing in this configuration, or if it does, the docs should maybe reflect that.

Funny thing - I have written my own version of a sync library, and by adding hooks that trigger on create/update (and sometimes remove a property from the object before it is added/updated in the database to indicate that this transaction should NOT be automatically synced) the update-hook seems to fix the issue above (making the pseudo-code I posted run smoothly), even when it does not actually do anything to modify the object and only creates a sync record in a different table. Which is why it was so extremely difficult to hunt down this bug, because in my previous code everything ran fine (where I used the sync hooks), and in any new version I attempted my code broke (because I commented out the hook function calls to first test the new code without the added complexity, haha). I have also stopped providing the optional parameter on "put" operations.

I'm not complaining about the bug; though my time at work could have been more productive than to spend many hours hunting down this bug, I am extremely grateful for dexie and only wanted to let you know about what seems to be a bug to help you improve this awesome implementation!

P.S.: The primary key being also part of a compound index is mostly for safety reasons. Such as for removing the entry with appKey X, but only if the entry belongs to the current user. Only read the entry with appKey X if the entry belongs to the current user. It doesn't prevent people from going into the IndexedDB and delete entries there (or attempt to decrypt them), but it does provide an extra safety net for the application, in case someone does accidentally screw up an url or something goes wrong. And I could check those things without a compound index, but that would be a lot more cumbersome, to basically re-do the work a compound index is supposed to do for me.

**Comment by dfahlander** [comment]
I know the feeling ;) Sorry for that. If the setup you describe ('++appKey, userId, [appKey+userId]') does not set the compound index correctly, that would be a browser bug. It would be easy to add this in a unit test to check what browsers that would be affected by the bug.

Regarding the issue with add(x, key) - it should fail (not crash, but return a rejected promise) if the primary key is [inbound](http://dexie.org/docs/inbound). On tables with inbound keys, one must not supply the optional key argument to either [put](http://dexie.org/docs/Table/Table.put()) nor [add](http://dexie.org/docs/Table/Table.add()). I was actually quite sure this was documented, but looking at the docs, it's not. The raw IndexedDB api [IDBObjectStore.put()](https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/put) and [IDBObjectStore.add()](https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/add) has the same rule - maybe a non-intuitive rule, but the logic behind is that the key should be given canonically within the object property instead for inbound keys.

When using hooks, the code-path in dexie differs as it has to understand the primary key in order to call hooks appropriately. That code may be more forgivable than the raw indexedDB API is. 

In summary, the things needed to do in dexie to go forward here, is to:
1. Test your setup (combining autoincremented primary key with compound index containing that key) and see at what browsers this fails. 
2. Update documentation of Table.put() and add() in dexie docs to clarify the optional key parameter.
3. Make sure we have a consistent way of treating the optional key argument to Table.add()/put() no matter if hooks are there or not. Either make hooked version complain, or (unlike the raw API) allow the optional parameter in non-hooked version. If so, need to overwrite the inbound property if a key was given.

**Comment by Fareley** [comment]
I did test it again, just to be sure. The latest version of chrome definitely does not set the compound index on the first add, only when I put the object again does it properly show. I'm fairly certain that the latest version of firefox experiences the same issue, as I was developing the app primarily with firefox to test and only when I had no idea why the compound index failed did I turn to chrome (where you can inspect the compound indexes manually). The app does not permit any other browser, so I can't test that easily and cannot justify re-writing all the restrictions and implementing tests for other browsers which we don't support at all on company time, to be honest :(.

As for the other two points - I have no idea how to do those things, to be honest, and don't really dare to try. But thank you for listening and caring :)

**Comment by dfahlander** [comment]
Sorry, the bullets were meant as a todo-list for this issue for any contributor including myself. I reopened the issue as I intend we do something about these three bullets.

**Comment by dfahlander** [comment]
I've reproduced this bug and hear and, Lo and behold, both Chrome and Firefox have this bug but not Safari. IE/Edge don't even support compound index.
